rust-mode.el

```diff
@@ -71,9 +71,15 @@ instead of `prog-mode'. This option requires emacs29+."
     map)
   "Keymap for Rust major mode.")
 
-(if (and (version<= "29.1" emacs-version) rust-mode-treesitter-derive)
-    (require 'rust-mode-treesitter)
-  (require 'rust-prog-mode))
+;;;###autoload
+(defun rust-mode ()
```

Comment:
Doesn't this function conflict with the `rust-mode` function that is exposed by the rust-prog-mode / rust-treesitter-mode ?

Reply:
Infact, give that you are calling `rust-mode` in the body of this function - it makes me more nervous. Do you think we can rename it to a different function instead, probably `rust-mode-load` or something like that ?

Reply:
@psibi yeah, it is a bit hacky...

The thing is that we want to autoload `rust-mode.el` when the user calls `(rust-mode)` and then select the actual major mode. However, both versions of the major mode (`rust-prog-mode` and `rust-treesitter-mode`) are called `rust-mode` and which one to choose depends on the environment (the emacs version and the value of `rust-mode-treesitter-derive`). So, I don't know how to autoload one or the other conditionally.

Also, we cannot change the name of the major modes (and autoload both) because the hooks are named after them and other extensions expect these modes to be called `rust-mode`.

So, this PR autoloads a stub called `rust-mode` that is redefined by the specific call to `define-derived-mode`. Thus, the following calls to `rust-mode` execute the `rust-mode` function corresponding to the actual major mode.

Reply:
> probably `rust-mode-load` or something like that ?

Or `rust-mode-maybe`, it's mentioned in the corresponding issue. But using another name may break users' config as many people manually do

```elisp
(add-to-list 'auto-mode-alist '("\\.rs\\'" . rust-mode))
```